degirum_tools/line_count.py

```diff
@@ -96,44 +121,97 @@ def analyze(self, result):
+                trail_onto_line_projection_sign = np.sign(trail_onto_line_projection)
+                if np.all(trail_onto_line_projection_sign == np.sign(line_vector)):
+                    counts.top += 1
+                else:
+                    if not np.any(trail_onto_line_projection_sign):
+                        if cross_product > 0:
+                            counts.bottom += 1
+                        else:
+                            counts.top += 1
+                    else:
+                        counts.bottom += 1
 
-        def count_increment(counts, trail_start, trail_end):
-            if trail_start[0] > trail_end[0]:
-                counts.left += 1
-            else:
-                counts.right += 1
-            if trail_start[1] > trail_end[1]:
-                counts.top += 1
-            else:
-                counts.bottom += 1
+        if not self._accumulate:
+            self._line_counts = [LineCounts() for _ in self._lines]
 
         for tid in new_trails:
             trail = get_anchor_coordinates(
                 np.array(result.trails[tid]), self._anchor_point
             )
             if len(trail) > 1:
-                trail_start = trail[0]
+                trail_start = trail[0] if self._whole_trail else trail[-2]
                 trail_end = trail[-1]
 
-                for total_count, line in zip(self._line_counts, self._lines):
-                    if intersect(line[:2], line[2:], trail_start, trail_end):
-                        self._counted_trails.add(tid)
-                        count_increment(total_count, trail_start, trail_end)
+                for total_count, line, line_vector in zip(
+                    self._line_counts, self._lines, self._line_vectors
+                ):
+                    if intersect(line[0], line[1], trail_start, trail_end):
+                        if self._count_first_crossing:
+                            self._counted_trails.add(tid)
+                        trail_vector = self._line_to_vector((trail_start, trail_end))
+                        cross_product = np.cross(trail_vector, line_vector)
+                        trail_onto_line_projection = self._projection(
+                            line_vector, trail_vector
+                        )
+                        count_increment(
+                            total_count,
+                            trail_vector,
+                            line_vector,
+                            cross_product,
+                            trail_onto_line_projection,
+                        )
                         if self._per_class_display:
                             class_count = total_count.for_class.setdefault(
                                 result.trail_classes[tid], SingleLineCounts()
                             )
-                            count_increment(class_count, trail_start, trail_end)
+                            count_increment(

```

Comment:
Now, when count_increment becomes complex function, it is not good to call it twice with the same set of input parameters. Propositions:
1. Implement `SingleLineCounts.__add__()` method to add two `SingleLineCounts` objects
2. Return `SingleLineCounts` object from `count_increment()`, containing only delta. Just create new empty one inside `count_increment()`, increment it, and return. Assing it to local var, say, `increment`
3. When you update class_count, just do this: `class_count = class_count + increment`